Enable RISC-V PMP support on RV32 QEMU

cwshu opened the pull request:
This PR tries to enable RV32 QEMU to improve CI of RISC-V PMP and userspace. It fixes the existing bugs and adds a new feature to support PMP on non-XIP RISC-V system (because RV32 QEMU is non-XIP).

Basically, this PR contains 4 parts of commits:

  1. [Commit 1 ~ 5] Bugfix of RISC-V PMP and userspace
  2. [Commit 6 ~ 10] Cleanup of RISC-V PMP and userspace. It's preparation to support ROM protection via PMP
  3. [Commit 11 ~ 12] Add new feature, ROM protection via PMP, to support non-XIP RISC-V system
  4. [Commit 13] Enable RISC-V PMP in RV32 QEMU

**Detail about commits**

The 2nd part, cleanup of RISC-V PMP and userspace is mainly for
   - Introduce `struct riscv_pmp_region` to modulize PMP CSR handling code and handle NAPOT/TOR mode more easily
   - Enhance MPU log for debugging use

The example of new MPU log of RISC-V PMP is like below:
```
D: PMP context 0x80011c08 add region 0: (0x80011034, 0x00000004, R  , NA4)
D: PMP context 0x80011c08 add region 1: (0x80014000, 0x00000800, RW , NAPOT)
D: PMP context 0x80011c08 add region 3: (0x80011000, 0x00000008, RW , NAPOT)
D: PMP context 0x80011c08 add region 4: (0x80010000, 0x00001000, RW , NAPOT)
D: Clear all dynamic PMP regions
D: Apply user PMP context 0x80011c08 to dynamic PMP regions
```

The 3nd part, ROM protection via PMP is

  - Protecting ROM (code/rodata) via static & locked PMP region in both Zephyr user & supervisor modes (RISC-V U/M-mode)
  - This is one of [Boot Time Memory Configuration](https://docs.zephyrproject.org/latest/reference/usermode/memory_domain.html#boot-time-memory-configuration) in Zephyr userspace support.

**Relation to other PRs**

  - Open another PR #39301 for gen_app_partitions script fix.
  - ~Part 2 has been already in #36534.~

Commits in other PRs will be removed here after they get merged.

Resolve 1st part of #31811

cwshu committed:
riscv: linker: fix duplicate symbol when CONFIG_USERSPACE=y

_image_ram_start symbol is duplicate in CONFIG_USERSPACE. This symbol
should be at the start of app_smem in userspace, so remove another
symbol in CONFIG_USERSPACE.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
soc: riscv: virt: enable RISC-V PMP support

Enable CONFIG_RISCV_PMP in qemu virt soc. Use this SoC as CI testing
platform of RISC-V PMP and Userspace.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: enable MPU log module for debugging

Cleanup logging API in core_pmp.c. Remove old printf-based debugging API
and change the log module of PMP to individual MPU log module.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: simplify thread initialization

Thread init related to PMP & userspace contains 5 parts:

1. User/supervisor thread clear PMP context
2. User thread clear it's context
3. User/supervisor thread assign to different entry
4. Supervisor thread assign mstatus.MPRV for M-mode PMP protection
5. User/supervisor thread setup PMP regions of stack guard if enabled

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: introduce riscv_pmp_region structure

Using struct riscv_pmp_region to modulize PMP CSR handling, including
PMP NAPOT/TOR mode handling. This patch can make us more easily to
add/remove RISC-V PMP regions without considering register handling.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: simplify pmp region number computation

Simplify multiple ifdef case in computing region number. Also move these
macros to core_pmp.c because they are only used in one file.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: reorder function definitions

Reorder the memory domain async functions to:
  arch_mem_domain_partition_add()
  arch_mem_domain_partition_remove()
  arch_mem_domain_thread_add()
  arch_mem_domain_thread_remove()

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: fix sp of supervisor thread in _Fault function.

Although CONFIG_USERSPACE is enabled, there are supervisor threads who
don't have privileged stack using exception handler. Only let user
threads to switch to privileged stack in exception handler.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
tests: disable division-by-zero test in RISC-V arch

Because RISC-V arch (M extension) doesn't trigger exception for
division-by-zero, this test can't support RISC-V and is disabled in
RISC-V.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: change mechanism of arch_buffer_validate()

Implement new mechanism of arch_buffer_validate() to support checking
static PMP regions. This is preparation patch for code/rodate protection
via RISC-V PMP.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: pmp: add PMP protection of code and rodata

This commit enable PMP-based memory protection of code and rodata
instead of relying on non-writable real HW (e.g. flash). Use static
PMP region with PMP Lock bit to protect them in both user/supervisor
mode.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu commented:
@fkokosinski Thank you for reviewing.
I have close #36534 to proceed all commits in this PR.

cwshu commented:
Specialize ROM protection via PMP region in XIP mode: protect whole flash area as ROM region instead of using rom_region symbols.
It is used in Cortex-M implementation: `soc/arm/common/cortex_m/arm_mpu_regions.c`
Try to resolve #39922

cwshu committed:
arch: riscv: pmp: fix return value of arch_mem_domain_partition_remove()

If no thread use this memory domain, there isn't any user PMP region
translated from memory partitions in domain. In this case, memory
partition removal doesn't need to remove user PMP region and
arch_mem_domain_partition_remove() could just successfully return.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu commented:
Rebase to main to fix merge conflict and use new CI.
Add new bugfix of `arch_mem_domain_partition_remove()` after commit 1cd7cccbb12d3f67e193e66a4c0b3de6bc982c2c

cwshu commented:
Remove ROM protection via PMP in XIP mode because this feature becomes larger and I create an individual PR #40647 for it.

cwshu commented:
@kgugala @pgielda @tgorochowik @fkokosinski @mgielda @katsuster @henrikbrixandersen 
Hi reviewers, my PR is ready for review. Thank you.

fkokosinski commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -582,24 +1064,18 @@ int arch_mem_domain_thread_remove(struct k_thread *thread)
-	uchar_pmpcfg = (unsigned char *) thread->arch.s_pmpcfg;
-
-	uchar_pmpcfg++;
+	struct riscv_pmp_region dynamic_regions[3]; /* Maximum region_num is 3 */
```
In commit 4478fbaa0130464410dadf183553f0afbbf14cce you introduce the maximum dynamic region number as 4, then you change it to 3. Why is that the case?

cwshu replied:
Commit c09593fa11555aaf8a9597c8364d9492ae3ec70a add static PMP regions, and the MPRV region (enable permission by default) should have lower priority than static regions.
So I manage MPRV region globally in `mprv_region` and remove it from dynamic PMP regions in `z_riscv_init_stack_guard()`

fkokosinski reviewed (commented):
After re-reviewing, I have the following questions. Looks good otherwise.

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -6,21 +6,144 @@
+# define PR_ADDR "0x%08lx"
+#elif __riscv_xlen == 64
+# define PR_ADDR "0x%016lx"
+#endif
```
Thank you. I forget this.
Done.

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -6,21 +6,144 @@
+#ifdef CONFIG_64BIT
+# define PMPCFG_NUM(index)	(((index) / 8) * 2)
+# define PMPCFG_SHIFT(index)	(((index) % 8) * 8)
+#else  /* CONFIG_64BIT */
```
May be `!CONFIG_64BIT`.
I remove the comments because it's trivial. (`#ifdef` is near to `#else`)
Done.

cwshu commented:
@fkokosinski 
Thanks a lot for reviewing this large patch.
I have fixed all questions.

enjiamai reviewed (changes requested):
Hi , @cwshu , thanks so much for the bug fixing and feature adding. I have some personal thoughts for this PR: 
The comments and some macro/struct you are adding for it are great, I am all for it. They make the code more clear and readable. I guess if we can make less change on this PR, or add things based on its original structure as possible, that would be easier for reviewers. I understand you want to add some common subroutines to clean up and make it more readable, and it should be. From my point of view, the original code makes a good code/size efficiency under different configurations, even it has lots of #ifdef. So I think the subroutines you made can be better designed/divided, at least there is no __unused existed.

I'm not good at Risc-V, if it's not right, correct me, please :)

cwshu commented on tests/ztest/error_hook/src/main.c:
```diff
@@ -267,7 +267,9 @@ void test_catch_fatal_error(void)
 #if defined(CONFIG_USERSPACE)
 	run_trigger_thread(ZTEST_CATCH_FATAL_ACCESS);
 	run_trigger_thread(ZTEST_CATCH_FATAL_ILLEAGAL_INSTRUCTION);
+#if !defined(CONFIG_RISCV)
```
Thanks, I will add comment later.

cwshu replied:
Fixed

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -6,21 +6,144 @@
+# define PMPCFG_SHIFT(index)	(((index) % 8) * 8)
+#else
+# define PMPCFG_NUM(index)	((index) / 4)
+# define PMPCFG_SHIFT(index)	(((index) % 4) * 8)
```
Compiler could optimize them to the bitwise operations, so I wrote them in the more readable way.

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -608,79 +1084,85 @@ void z_riscv_init_stack_guard(struct k_thread *thread)
+	memset(thread->arch.s_pmpcfg, 0,
+		sizeof(ulong_t) * PMP_CFG_CSR_NUM_FOR_STACK_GUARD);
+	memset(thread->arch.s_pmpaddr, 0,
+		sizeof(ulong_t) * PMP_REGION_NUM_FOR_STACK_GUARD);
```
OK, I will rollback to previous one.

cwshu replied:
Fixed

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -143,149 +266,448 @@ void csr_write_enum(int pmp_csr_enum, ulong_t value)
 }
 
-int pmp_get(unsigned int index, ulong_t *cfg_val, ulong_t *addr_val)
+static __unused int riscv_pmp_get(unsigned int index, uint8_t *cfg_val, ulong_t *addr_val)
```
Fixed

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -143,149 +266,448 @@ void csr_write_enum(int pmp_csr_enum, ulong_t value)
 
-/* Function to help debug */
-void z_riscv_pmp_print(unsigned int index)
+static __unused int riscv_pmp_regions_set_from_last(int last_reg_index,
```
Fixed

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -143,149 +266,448 @@ void csr_write_enum(int pmp_csr_enum, ulong_t value)
+	return reg_index;
+}
+
+static __unused int riscv_pmp_region_translate(ulong_t pmpcfg[], ulong_t pmpaddr[], int index,
```
Fixed

cwshu commented on arch/riscv/core/pmp/core_pmp.c:
```diff
@@ -143,149 +266,448 @@ void csr_write_enum(int pmp_csr_enum, ulong_t value)
-#if defined(CONFIG_USERSPACE)
-#include <linker/linker-defs.h>
-void z_riscv_init_user_accesses(struct k_thread *thread)
+static __unused int riscv_pmp_regions_translate(ulong_t pmpcfg[], ulong_t pmpaddr[],
```
Fixed

cwshu commented:
@enjiamai
Thanks for reviewing this patch!

I still hesitate to split bugfix, cleanup code, feature code, and QEMU enablement code into different PRs or merge them in a single PR. The only reason I merge them in a single PR is because it's 1st time to enable QEMU for RISC-V PMP. I need a QEMU platform to let community know these commits are well-tested and verified.

It is not easy to add feature of ROM protection (part 3) without PMP cleanup code (part 2).
(Also, adding feature in #40647 bugfix w/o PMP cleanup code is difficult.)
Adding PMP support of non-XIP platform `qemu_riscv32` (part 4) need feature of ROM protection, so I need all these parts of code.

I have removed all `__unused` macros and merged `riscv_pmp_region_[set|translate]()` functions to make code cleaner.

enjiamai reviewed (approved):
It look good for me, and I verified #40647 it solved the problem. Thanks for the work!

dcpleung reviewed (approved):
LGTM, but I am no expert on RISC-V.

cwshu commented:
Sorry, This PR should get merged later.
@fkokosinski find a bug when enabling `CONFIG_PMP_STACK_GUARD=y` in #40647, I will try to fix it soon.

cwshu committed:
arch: riscv: pmp: fix IRQ handling of PMP stack guard

This commit add 2 minor fixes of IRQ handling:

1. Save caller registers before calling z_riscv_configure_stack_guard()
in RISC-V assembly.

2. reschedule and no_reschdule code paths use different interrupt
return path after supporting of CONFIG_PMP_STACK_GUARD. back-to-back
interrupt checking is in the reschedule code path so that it should
jump to interrupt return path of reschedule.

Signed-off-by: Jim Shu <cwshu09@gmail.com>

cwshu commented:
Add new commit de98246ae9595ebfec59c7c1e8daa5a8bf64cc94 to fix `CONFIG_PMP_STACK_GUARD=y` bug.
Remove DNM label as the bug has been fixed.

cwshu commented:
It looks like I don't have permission to edit label.
Could anyone help me to remove DNM label?
